### ISSUE TITLE
Objective evaluation via SWE-bench 

### ISSUE THREAD

**Issue by ai-christianson** [ai-christianson/RA.Aid#12]
Right now, our prompts and agent process are hand tuned and evaluated manually and subjectively.

Without objective evaluation criteria, it's very hard to tell if a prompt tweak or other logic change has made a true improvement and hadn't caused regressions in performance. 

SWE-bench (https://www.swebench.com/) is an evaluation framework designed specifically for evaluating SWE agents and therefore would be a great first target. 

We can start with a subset at first to get it working, then expand to run the full suite. 

Ideally this would be run in a GitHub action so we can track performance over time, detect regressions, and test multiple configurations (different models, expert models, web research enabled or not, etc.)

**Comment by ai-christianson** [comment]
I did some research on SWE-bench to understand exactly how we can use it to evaluate RA.Aid objectively. Here's what I found and what I think we should do:

## The Data
SWE-bench offers three dataset variants:
- Full dataset: 2,294 issue-commit pairs
- Lite: 300 optimized test instances 
- Verified: 500 human-annotated problems

I recommend we start with SWE-bench Lite because:
1. It's specifically optimized for reduced computational costs
2. Perfect for initial testing and integration
3. Still gives us meaningful metrics (the best agents achieve ~43% success rate on Lite)

## How Testing Works
The testing process has two phases:
1. **Inference**: Feed the agent a GitHub issue and let it generate a fix
2. **Evaluation**: Test the fix using the project's test suite

Key things I discovered about the evaluation process:
- Test cases and unit tests are publicly available
- No restrictions on using test results during development
- TDD and iterative development are explicitly allowed
- Focus is on final performance rather than methodology

This is good for us because RA.Aid already does iterative development with test feedback - we don't need to modify its behavior at all.

## Proposed Integration Plan

After analyzing the requirements, I think we should take this approach:

1. **Create a Dataset Generation Script**
   - Takes each SWE-bench Lite test case
   - Sets up repo at correct commit
   - Runs RA.Aid with the issue description
   - Captures the resulting diff
   - Formats it as required

2. **Output Format Requirements**
Each test case needs to generate an entry in all_preds.jsonl like this:
```json
{
    "instance_id": "owner__repo-PR-123",
    "model_patch": "diff --git a/file.py b/file.py\n..."
}
```

The directory structure needs to be:
```
evaluation/lite/YYYYMMDD_raaid/
├── all_preds.jsonl
└── logs/
```

3. **First Phase: Single Test**
Run one test to verify our integration:
```bash
python -m swebench.harness.run_evaluation \
    --predictions_path evaluation/lite/YYYYMMDD_raaid \
    --max_workers 1 \
    --instance_ids [specific_id]
```

4. **Second Phase: Full Lite Dataset**
Once single test works, expand to all 300 Lite instances:
```bash
python -m swebench.harness.run_evaluation \
    --predictions_path evaluation/lite/YYYYMMDD_raaid \
    --max_workers 1
```

## Important Technical Notes
- Environment requires:
  - Python 3.8+
  - Docker installation
  - Dependencies: `pip install swebench datasets`
- Each test runs in its own Docker container
- Failed attempts should be omitted from predictions file
- Patches must apply cleanly to base commit
- Reasoning traces now required for leaderboard submissions

## Performance Context
- Top agents: ~20% success on full dataset
- Best performance on Lite: 43% success rate
- LLMs alone: 4.80% success rate on complex tasks

## Next Steps
1. Create the dataset generation script
2. Test with a single instance from Lite
3. Add logging for failed attempts
4. Expand to full Lite dataset (300 instances)
5. Analyze results and iterate on RA.Aid as needed

## Future Work
Once we have basic evaluation working:
1. Add GitHub Actions integration
2. Track performance over time
3. Test different configurations:
   - Different models
   - Different expert models
   - Web research enabled/disabled
4. Consider expanding to full dataset (2,294 instances)

**Comment by mrbar42** [comment]
tl;dr - add eval results to README

i'm not sure what would be the cost of running an entire eval but each ran result can be aggregated into a table in the README even if not all the results are going to be reported to SWE Bench itself.

having a table of SWE Bench results per model combination would be instrumental in showing the capability of this (awesome) tool as well as contributing yet another assessment of models performance as they are released like the Aider leaderboard.
Why would i use $60/M model when i can get 80% of the result at $0.28/M? aggregating data will show that.

**Comment by ariel-frischer** [comment]
I'm testing this out, I went with using [aider-swe-bench](https://github.com/Aider-AI/aider-swe-bench) as a template for running RA.Aid with swe-bench. I'm working on running this locally, and I am generating diffs for each task which can be used for the predictions. Did not integrate aider's `test_cmd` with `RA.Aid` yet - not even sure if that a configurable option yet. I can see how this can get quite costly for me even with the swe-bench-lite dataset. Will probably upload my implementation for public usage within next few days.

**Comment by ai-christianson** [comment]
@ariel-frischer I have some preliminary code at `https://github.com/ai-christianson/RA.Aid/blob/master/scripts/generate_swebench_dataset.py` but I have limited time to fully get this working.

If you're able to get it working (the script I started, the aider template, or anything), I'm willing to spend some $$ on the API calls to run the eval.

**Comment by ariel-frischer** [comment]
> @ariel-frischer I have some preliminary code at `https://github.com/ai-christianson/RA.Aid/blob/master/scripts/generate_swebench_dataset.py` but I have limited time to fully get this working.
> 
> If you're able to get it working (the script I started, the aider template, or anything), I'm willing to spend some $$ on the API calls to run the eval.

Nice, give me couple minutes I'm going to post my repo on Github. I got most of the prediction generation code running but need to dive into evaluations more, as the original aider-swe-bench uses an older swe-bench version that no longer uses the same library methods. I've been sick last couple days so haven't been able to finish up the rest but I'll continue digging into it. This is the code I'm working on refactoring: [](https://github.com/Aider-AI/aider-swe-bench/blob/main/report.py).

**Comment by ariel-frischer** [comment]
Okay here it is: https://github.com/ariel-frischer/RA.Aid-swe-bench
Hopefully installation with poetry works, there were some tough dependency issues I was able to resolve.
README.md is updated with current progress, and improvements/issues.

**Comment by ariel-frischer** [comment]
Your implementation is much better for generating predictions probably as I'm not doing these three things: 
- Creates or reuses a dedicated Python virtual environment via `uv venv`
- Installs `ra-aid` in editable mode + any project dependencies via `uv pip`
- Also installs the cloned project itself in editable mode if it appears to be a Python package

**Comment by ariel-frischer** [comment]
Ok, fixed an issue where I was only running `run_research_agent` without `run_planning_agent` because sometimes research agent would make code changes. I was really confused how state gets shared from research agent to planning agent but I see now how it goes through `_global_memory` store. Now that I'm running research agent (research-only mode) then planning agent, I am getting more reliable diffs after each attempt. I also fixed a bug with 1 MAX_THREAD. Also fixed os working directory errors with context manager.

**Comment by ariel-frischer** [comment]
@ai-christianson Going to continue improving my repo and integrating a few things you've set up. Just now realizing all the dataset repos are python, this makes setup much simpler. Few things I have in mind:
- Integrate virtual env + install deps
- Add failing/passing tests to prompt
- Finally fix up report.py for grading/evaluation
- Clean cli/tui is a nice to have extra feature

**Comment by ai-christianson** [comment]
Excellent! I will try it out when you're ready.

**Comment by ariel-frischer** [comment]
Update:
- [X] Integrate virtual env + install deps
- [X] Add failing/passing tests to prompt
- [ ] (NEW) Extracting and automating `test_cmd` for improved retries + winner selection.
- [ ] Finally fix up report.py for grading/evaluation
- [ ] Clean cli/tui is a nice to have extra feature

Got a nice caching RepoManager class now. Using `git worktree add` + symbolinks for .venv handling. The `uv` installer method is coming in clutch.

**Comment by ariel-frischer** [comment]
Update:

- [X] Integrate virtual env + install deps
- [X] Add failing/passing tests to prompt (Removed in SUBMIT_MODE due to submission rules)
- [X] Added SUBMSSION_MODE to qualify for submission by following the [SWE checklist](https://github.com/swe-bench/experiments/blob/main/checklist.md).
- [X] Add max TIMEOUT for each ra-aid process
- [X] Added trajectory output files per submission rules.
- [ ] Need some fix for inner aider streaming readability issues in trajectory files.
- [ ] ~~(NEW) Extracting and automating `test_cmd` for improved retries + winner selection.~~ (Removed in SUBMIT_MODE due to submission rules)
- [ ] Finally fix up report.py for grading/evaluation
- [ ] Clean cli/tui is a nice to have extra feature

**Comment by ariel-frischer** [comment]
Ok, I would say the project is ready in general. We have working prediction generations with `make run` and should have solid evaluation results with `make eval`. I was able to strip out broken chunks of legacy code in report.py, and refactor it for readability. Most of it is borked. We do automatically track `evaluated` prediction files so that we don't evaluate the same files twice. We have `make reset-eval` to reset prediction file evaluation fields if you want. Its working for me does not have all the bells and whistles yet but should give us submission quality predictions. 

Not sure how to approach this one:
- [ ] Need some fix for inner aider streaming readability issues in trajectory files.

Will add .env file support soon. The Makefile seems simple enough for me to not dive into some full cli wizard.
Let me know if you have any questions/concerns/bugs. Also curious on evaluation results for top models! I've confirmed Deepseek v3 can actually resolve some of these successfully.

**Comment by ariel-frischer** [comment]
Continued with many useful updates. I'm quite happy with the setup now, perhaps there is one last bug/issue with venv setup on the `matplot` repo.

Regardless wanted to share some promising eval results for a small sample set of instances:
```
    "total_instances": 300,
    "submitted_instances": 13,
    "completed_instances": 9,
    "resolved_instances": 7,
    "unresolved_instances": 2,
    "empty_patch_instances": 4,
    "error_instances": 0,
    "completed_ids": [
        "django__django-11099",
        "django__django-12708",
        "django__django-13551",
        "django__django-14017",
        "django__django-14238",
        "mwaskom__seaborn-3190",
        "pytest-dev__pytest-9359",
        "sympy__sympy-15678",
        "sympy__sympy-21379"
    ],
    ...
```
Which is `53.8%` I could only achieve this with anthropic `claude-3-5-sonnet-20241022`. It is much better than Deepseek V3 in this langgraph system. It actually runs shell commands and tests, I haven't seen Deepseek run any of those perhaps because that model is not yet supported.

**Comment by jimwhite** [comment]
I've been working on getting `aider-swe-bench` running in a Colab notebook (to make a workflow for the K-prize https://www.kaggle.com/competitions/konwinski-prize) and having a lot of version compatibility issue with Aider (currently stuck at 0.40.6).  This is awesome work and I'm going to switch to RA.Aid.  Disappointing to hear about DeekSeek V3 not working well though, that has actually been a positive aspect of my trials with Aider.

**Comment by ariel-frischer** [comment]
> I've been working on getting `aider-swe-bench` running in a Colab [notebook](https://github.com/fovi-llc/aider-swe-bench/blob/main/notebooks/aider_swe_bench_colab.ipynb) (to make a workflow for the K-prize https://www.kaggle.com/competitions/konwinski-prize) and having a lot of version compatibility issue with Aider (currently stuck at 0.40.6). This is awesome work and I'm going to switch to RA.Aid. Disappointing to hear about DeekSeek V3 not working well though, that has actually been a positive aspect of my trials with Aider.

Yea, I wish it worked with Deepseek V3 as well given the huge cost savings. I notice you can configure `ra-aid` to use one model for the langgraph agents with `--model` and set the aider editor model with ENV variable: `AIDER_MODEL`. Should save a bit while editing code.

**Comment by ai-christianson** [comment]
I'm currently away from home but will check this out more in depth when we get back. Awesome work on this! 

I think dsv3 can be made to work better but it will take some tuning of the prompts. 

Now that we have this eval, we'll actually be able to measure whether prompt changes are working or not.

I think running eval on qwen 32b coder instruct is worthwhile too.

**Comment by jimwhite** [comment]
> I think dsv3 can be made to work better but it will take some tuning of the prompts.
> 
> Now that we have this eval, we'll actually be able to measure whether prompt changes are working or not.

Indeed.  That's why I'm working on the K-prize which requires open models.  My plan is to build a coding agent using DSPy (or similar) RL framework to do prompt tuning automagically.

https://dspy.ai/

**Comment by ai-christianson** [comment]
Cool idea. I can see how automated eval would help out with automatic prompt optimization.

**Comment by ariel-frischer** [comment]
Some more promising results on another small batch with Sonnet 3.5:

```
    "total_instances": 300,
    "submitted_instances": 17,
    "completed_instances": 15,
    "resolved_instances": 10,
    "unresolved_instances": 5,
    "empty_patch_instances": 2,
    "error_instances": 0,
    "completed_ids": [
        "django__django-10924",
        "django__django-11099",
        "django__django-11283",
        "django__django-12184",
        "django__django-13321",
        "django__django-13768",
        "django__django-13925",
        "django__django-14155",
        "django__django-14411",
        "django__django-15695",
        "django__django-16527",
        "psf__requests-3362",
        "sphinx-doc__sphinx-8713",
        "sympy__sympy-21627",
        "sympy__sympy-22005"
    ],
```

So `10/17 = 0.58823` or `58.8%` is pretty solid. I was able to fix a significant environment setup issue: https://github.com/ariel-frischer/RA.Aid-swe-bench/issues/4 which improved `ra-aid` shell commands.

**Comment by ai-christianson** [comment]
I got your swe-bench code running. I'm trying to reproduce your results. My agent got stuck in a loop:

```
╭─────────────────────────────────────────────────────────────────────────────────── Error ────────────────────────────────────────────────────────────────────────────────────╮
│ Encountered BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000           │
│ maximum'}}. Retrying in 4s... (Attempt 3/20)                                                                                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
2025-01-21 13:09:05,959 - ra_aid.ra_aid.agent_utils - WARNING - API error (attempt 4/20): Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000 maximum'}}
╭─────────────────────────────────────────────────────────────────────────────────── Error ────────────────────────────────────────────────────────────────────────────────────╮
│ Encountered BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000           │
│ maximum'}}. Retrying in 8s... (Attempt 4/20)                                                                                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
2025-01-21 13:09:17,656 - ra_aid.ra_aid.agent_utils - WARNING - API error (attempt 5/20): Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000 maximum'}}
╭─────────────────────────────────────────────────────────────────────────────────── Error ────────────────────────────────────────────────────────────────────────────────────╮
│ Encountered BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000           │
│ maximum'}}. Retrying in 16s... (Attempt 5/20)                                                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
2025-01-21 13:09:35,754 - ra_aid.ra_aid.agent_utils - WARNING - API error (attempt 6/20): Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000 maximum'}}
╭─────────────────────────────────────────────────────────────────────────────────── Error ────────────────────────────────────────────────────────────────────────────────────╮
│ Encountered BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000           │
│ maximum'}}. Retrying in 32s... (Attempt 6/20)                                                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
2025-01-21 13:10:09,061 - ra_aid.ra_aid.agent_utils - WARNING - API error (attempt 7/20): Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000 maximum'}}
╭─────────────────────────────────────────────────────────────────────────────────── Error ────────────────────────────────────────────────────────────────────────────────────╮
│ Encountered BadRequestError: Error code: 400 - {'type': 'error', 'error': {'type': 'invalid_request_error', 'message': 'prompt is too long: 202682 tokens > 200000           │
│ maximum'}}. Retrying in 64s... (Attempt 7/20)                                                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

Is there a good way to run the eval with a local version of ra-aid, e.g. `pip install -e`?

**Comment by ariel-frischer** [comment]
There might be a way to somehow do that with `pyproject.toml` with localpath config?. I've set it to `dev` version remotely with: `ra-aid = { git = "https://github.com/ariel-frischer/RA.Aid.git", branch = "dev", rev = "f331f041d3ea18239cdcbbdedce39af65a02e26d" }
`
But if you just delete `ra-aid` dependency, remove .venv, reinstall, I would imagine it should find `ra-aid` cli from your shell PATH and use that to point to your local `ra-aid` executable.

**Comment by ai-christianson** [comment]
This seems to have worked in `pyproject.toml`:

```
ra-aid = { path = "/home/user/workspace/ra-aid", develop = true }
```

**Comment by ai-christianson** [comment]
On latest commit as of today it's still getting stuck in a context too long error loop.

That said, the eval work you did looks excellent and will be extremely useful going forward.

**Comment by ariel-frischer** [comment]
> On latest commit as of today it's still getting stuck in a context too long error loop.
> 
> That said, the eval work you did looks excellent and will be extremely useful going forward.

With sonnet3.5 right?

Darn could be because we estimate tokens rather then getting the exact value. But I think more likely is that tool calls are calculated as 0 tokens because they are lists, maybe we need to convert them to strings first.

That state modifier was quite tricky to implement. By default, if you pass the anthropic model as the token counter it breaks due to some tool use error but that is the way it's shown in the docs. Binding the tools to the model did not seem to work either. Also using the ChatAnthropic model as the token counter param does an API request. I thought local token count estimates would also be faster.

**Comment by ai-christianson** [comment]
Yeah sonnet 3.5.

**Comment by ai-christianson** [comment]
Good news:

After setting our token estimation to be **3** tokens per byte (it was 4), I was able to run a small batch of predictions and run evals:

![Image](https://github.com/user-attachments/assets/deda088a-f6af-438c-a700-0d124d9690f4)

![Image](https://github.com/user-attachments/assets/c3326e61-5f00-4e7f-a25c-f6e4bdb20910)

![Image](https://github.com/user-attachments/assets/a5d6162a-ba5f-4913-87f2-dd3aa4872fd6)

**Comment by ai-christianson** [comment]
Going to close this as we have evals working and can open issues in the appropriate repos for incremental improvements.
